hs-abci-sdk/src/Tendermint/SDK/Modules/Validators/EndBlock.hs

```diff
@@ -0,0 +1,50 @@
+module Tendermint.SDK.Modules.Validators.EndBlock where
+
+import           Control.Monad                            (foldM)
+import qualified Data.Map.Strict                          as Map
+import qualified Data.Set                                 as Set
+import qualified Network.ABCI.Types.Messages.FieldTypes   as ABCI
+import qualified Network.ABCI.Types.Messages.Request      as Request
+import           Polysemy                                 (Members, Sem)
+import           Tendermint.SDK.BaseApp                   (BlockEffs,
+                                                           EndBlockResult (..))
+import qualified Tendermint.SDK.BaseApp.Store.List        as L
+import qualified Tendermint.SDK.BaseApp.Store.Map         as M
+import qualified Tendermint.SDK.BaseApp.Store.Var         as V
+import           Tendermint.SDK.Modules.Validators.Keeper
+import           Tendermint.SDK.Modules.Validators.Store
+import           Tendermint.SDK.Modules.Validators.Types
+
+
+endBlock
+  :: Members BlockEffs r
+  => Members ValidatorsEffs r
+  => Request.EndBlock
+  -> Sem r EndBlockResult
+endBlock _ = do
+  updatesMap <- getQueuedUpdates
+  curValKeySet <- getValidatorsKeys
+
+  -- update the Validators map and key set
+  newValKeySet <- foldM (\cvks (key, newPower) ->
```

Comment:
I think it would be better to implement this with a normal fold in `StateT (Set PubKey_) (Sem r) ()` rather than `foldM` because in the end you are doing a lot of extra work concatenating all of those sets at the end rather than single targeted updates.

Reply:
I'm not really following you here. I'm not very familiar with StateT, but i get the idea that it can be used in place of a foldM (sorta). If the only issue is not wanting to do the repeated Set.insert / Set.delete, would something like this be acceptable?

```
  (adds, rems) <- foldM (\(a,r) (key, newPower) ->
    if newPower == 0 then do
       M.delete key validatorsMap
       return (a, key : r) 
    else do
       M.insert key newPower validatorsMap
       return (key : a, r)
    ) ([], []) (Map.toList updatesMap)

  let newValKeySet = Set.union (curValKeySet Set.\\ Set.fromList rems) (Set.fromList adds)
```
Which would just accumulate the items to be added/removed through list building and then do two Set operations at the end.

Reply:
I find it kind of interesting that you can hack polysemy but aren't really starting with a knowledge of `mtl`, not everyone's haskell trajectory goes like that 👏 


with `StateT` it would look like 

```haskell
  newValKeySet <- flip execStateT curValKeySet $
    forM_ (Map.assocs updatesMap) $ (key, newPower) ->
      if newPower == 0 
        then do
          -- delete from Validators map and key set
          lift $ M.delete key validatorsMap
          modify $ Set.delete key
        else do
          -- update power in Validators map and ensure key is in key set
          lift $ M.insert key newPower validatorsMap
          modify $ Set.insert key
```

where `lift` is from `Control.Monad.Trans.Class`. It's equivalent to a `fold`, but allows for interleaving other effects as well.